File: solutions/2026-09-09/cpp/2026-09-09-count-commas-in-range-ii.cpp
```cpp
// Problem: Count Commas in Range II
// Link: https://leetcode.com/problems/count-commas-in-range-ii/
// Approach:
// The problem asks us to count the total number of commas used when writing all integers from 1 to n.
// Commas are inserted every three digits from the right, starting from numbers with 4 or more digits.
//
// We can solve this by calculating the number of commas up to 'n' and subtracting the number of commas up to '0' (which is 0).
// The core idea is to count commas for numbers in specific digit length ranges.
//
// Let's define a function `countCommasUpTo(x)` that counts the total commas for numbers from 1 to x.
//
// For a given number `x`, we can determine the number of digits it has. Let's say `x` has `D` digits.
//
// The numbers from 1 to 999 have no commas.
//
// Numbers from 1000 to 9999 (4 digits):
// - There are 9000 such numbers.
// - Each of these numbers has one comma (e.g., 1,000).
// - The first comma appears at the 1000th number.
//
// Numbers from 10000 to 99999 (5 digits):
// - There are 90000 such numbers.
// - Each of these numbers has one comma (e.g., 10,000).
//
// Numbers from 100000 to 999999 (6 digits):
// - There are 900000 such numbers.
// - Each of these numbers has one comma (e.g., 100,000).
//
// Numbers from 1000000 to 9999999 (7 digits):
// - There are 9000000 such numbers.
// - Each of these numbers has two commas (e.g., 1,000,000).
//
// Generalizing for a number `x` with `D` digits:
//
// We can iterate through powers of 10 to count commas.
//
// Let's consider the ranges based on powers of 1000 (which correspond to comma placements):
//
// 1-999: 0 commas
// 1000-999999: 1 comma (e.g., 1,000 to 999,999)
// 1000000-999999999: 2 commas (e.g., 1,000,000 to 999,999,999)
//
// For a number `x`:
// - If `x` has `D` digits.
// - The number of full groups of 3 digits from the right is `floor(D / 3)`.
// - Each number with at least `k` groups of 3 digits (i.e., numbers >= 10^(3*k)) will have at least `k` commas.
//
// We can calculate the total commas by considering the contribution of each comma position.
//
// Let's count commas for numbers from 1 to `N`.
//
// Consider the first comma position (thousands place): It appears in numbers from 1000 to N.
// The count of numbers from 1000 to N is `max(0LL, N - 999)`.
//
// Consider the second comma position (millions place): It appears in numbers from 1000000 to N.
// The count of numbers from 1000000 to N is `max(0LL, N - 999999)`.
//
// Consider the third comma position (billions place): It appears in numbers from 1000000000 to N.
// The count of numbers from 1000000000 to N is `max(0LL, N - 999999999)`.
//
// And so on.
//
// We can iterate through powers of 1000 (10^3, 10^6, 10^9, ...) up to a value less than or equal to `N`.
// For each power `p = 1000^k`, the numbers from `p` to `N` will have at least `k` commas.
// The numbers that contribute to the `k`-th comma (from the right) are those that are greater than or equal to `10^(3*k)`.
//
// Let `countCommas(num)` be the function to calculate commas up to `num`.
//
// For `num`:
// - If `num < 1000`, return 0.
// - Initialize `totalCommas = 0`.
// - Initialize `powerOf1000 = 1000`.
// - Initialize `commaCount = 1`.
// - While `powerOf1000 <= num`:
//     - The numbers from `powerOf1000` to `num` will have at least `commaCount` commas.
//     - We are interested in the count of numbers that have a comma at the `commaCount`-th position.
//     - The numbers that have the `commaCount`-th comma inserted are those >= `powerOf1000`.
//     - The count of such numbers up to `num` is `max(0LL, num - powerOf1000 + 1)`.
//     - Wait, this is not quite right. This counts all numbers that are >= `powerOf1000`.
//
// Let's rephrase: How many numbers from 1 to `num` have at least one comma? These are numbers >= 1000.
// Count is `num - 999` (if `num >= 1000`).
// How many numbers from 1 to `num` have at least two commas? These are numbers >= 1,000,000.
// Count is `num - 999999` (if `num >= 1000000`).
//
// General form: For the k-th comma (from the right), it appears for numbers >= `10^(3*k)`.
// The number of such integers from 1 to `num` is `max(0LL, num - (10^(3*k) - 1))`.
//
// So, the total number of commas is the sum of `max(0LL, num - (10^(3*k) - 1))` for `k = 1, 2, 3, ...` as long as `10^(3*k) <= num`.
//
// We need to be careful about integer overflow. `n` can be up to 10^15.
// `10^(3*k)` can reach up to `10^15`. Using `long long` is essential.
//
// Example: n = 1002
// k = 1: 10^(3*1) = 1000.
//   Count for k=1: `max(0LL, 1002 - (1000 - 1)) = max(0LL, 1002 - 999) = max(0LL, 3) = 3`.
//   Numbers are 1000, 1001, 1002. Each has one comma. Total 3.
// k = 2: 10^(3*2) = 1,000,000.
//   1,000,000 > 1002, so we stop.
// Total commas = 3.
//
// Example: n = 10000
// k = 1: 1000.
//   Count for k=1: `max(0LL, 10000 - 999) = 9001`.
//   This means there are 9001 numbers from 1000 to 10000 that have at least one comma.
// k = 2: 1,000,000.
//   1,000,000 > 10000, stop.
// Total commas = 9001.
// Let's check for n = 10000 manually:
// 1000 to 9999: 9000 numbers, each with 1 comma. Total 9000.
// 10000: 1 comma.
// Total = 9000 + 1 = 9001. Correct.
//
// The logic seems to be:
// `countCommas(num)` function:
//   Initialize `totalCommas = 0`.
//   Initialize `powerOf1000 = 1000`.
//   Initialize `threshold = 999`. (This is 1000 - 1)
//   While `powerOf1000 <= num`:
//     `count = num - threshold`; // Number of integers from `powerOf1000` to `num`
//     If `count > 0`:
//       `totalCommas += count`;
//     `powerOf1000 *= 1000`;
//     `threshold = threshold * 1000 + 999`; // This is the next `powerOf1000 - 1`
//     // Careful with overflow for threshold.
//     // A better way for threshold:
//     // `threshold = powerOf1000 - 1;`
//     // but this also has potential overflow for `powerOf1000`.
//     // Let's manage powers of 10 directly.
//
// Revised `countCommas(num)` function:
//   Initialize `totalCommas = 0`.
//   Initialize `powerOf3Digits = 1000LL`. // 10^3
//   Initialize `count = 1`. // Corresponds to the first comma position.
//   While `powerOf3Digits <= num`:
//     // Number of integers from `powerOf3Digits` to `num` that have a comma at this position.
//     // This is `num - powerOf3Digits + 1`.
//     // However, we need to consider the numbers that fall into this *group*.
//     // For example, for 1000 to 999999 (6 digits), there is one comma.
//     // The range for the first comma is 1000-9999.
//     // The range for the second comma is 1000000-999999999.
//
// Let's consider the contribution of each *digit position* from the right that is a comma position.
//
// For a number `N`:
//
// Number of digits in `N` is `D`.
//
// Commas occur at positions 3, 6, 9, 12, 15, ... from the right (1-indexed).
//
// For a given `N`, the number of times the 3rd digit from the right is a separator (i.e., the number has >= 4 digits) is `N - 999` if `N >= 1000`.
// These are numbers like `1,XXX`, `2,XXX`, ..., `N` (if N is 4 digits).
//
// The number of times the 6th digit from the right is a separator (i.e., the number has >= 7 digits) is `N - 999999` if `N >= 1000000`.
// These are numbers like `1,XXX,XXX`, `2,XXX,XXX`, ..., `N` (if N is 7 digits).
//
// Let's define `countCommas(num)`:
//
// If `num < 1000`, return 0.
//
// `totalCommas = 0`
//
// `// Count for numbers with one comma (e.g., 1,XXX)`
// `// These are numbers from 1000 up to num.`
// `// The range is [1000, num].`
// `// The number of integers is num - 1000 + 1.`
// `// If num < 1000, this count is 0.`
// `// The count of numbers from 1000 to num is max(0LL, num - 1000 + 1)`
// `// This counts the first comma for numbers >= 1000.`
// `// Let's think about the range of numbers for which a specific comma exists.`
//
// For numbers `1` to `N`:
//
// Count of numbers with at least one comma: `N >= 1000`. Count = `max(0LL, N - 999)`.
// These numbers contribute their first comma.
//
// Count of numbers with at least two commas: `N >= 1000000`. Count = `max(0LL, N - 999999)`.
// These numbers contribute their second comma.
//
// Count of numbers with at least three commas: `N >= 1000000000`. Count = `max(0LL, N - 999999999)`.
// These numbers contribute their third comma.
//
// This seems correct.
//
// The function `countCommas(num)`:
//   Initialize `totalCommas = 0`.
//   Initialize `limit = 1000LL`.
//   Initialize `separator = 999LL`.
//   Loop for `k = 1, 2, 3, ...`
//     If `limit > num`, break.
//     // The number of integers from `limit` to `num` is `num - limit + 1`.
//     // This is incorrect. The number of integers from `limit` to `num` is `num - limit + 1`.
//     // For the k-th comma (which appears for numbers >= 10^(3k)), the count of numbers up to N is N - (10^(3k) - 1) if N >= 10^(3k).
//     // Which simplifies to `max(0LL, N - (powerOf1000 - 1))`
//
// Let's try this structure:
// `long long solve(long long n)`:
//   `long long total_commas = 0;`
//   `long long power_of_1000 = 1000;`
//   `long long threshold = 999;` // Represents 1000 - 1, 1000000 - 1, etc.
//
//   // Loop as long as the current power of 1000 is less than or equal to n.
//   // This means we are considering comma positions that could exist in numbers up to n.
//   while (power_of_1000 <= n) {
//     // `threshold` is the number just before the `power_of_1000` mark.
//     // For example, when `power_of_1000` is 1000, `threshold` is 999.
//     // The count of numbers that have this comma inserted is `n - threshold`.
//     // These are numbers from `power_of_1000` up to `n`.
//     // Example: n=1002. power_of_1000=1000, threshold=999.
//     // Count = 1002 - 999 = 3. (Numbers are 1000, 1001, 1002)
//     // These 3 numbers each have their first comma.
//     // So, we add `n - threshold` to `total_commas`.
//     // We use `max(0LL, n - threshold)` in case n is smaller than threshold, though the while condition `power_of_1000 <= n` should prevent this for the first iteration.
//     // However, for subsequent iterations, it's possible that `power_of_1000` becomes very large, and `n - threshold` could be negative if `n` is smaller than `threshold`.
//     // The correct way to calculate the count of numbers between `power_of_1000` and `n` (inclusive) is `n - power_of_1000 + 1`.
//     // Let's rethink the meaning of `n - threshold`.
//     // If `n = 1002`:
//     // k=1: power_of_1000 = 1000, threshold = 999.
//     // Numbers from 1000 to 1002 have a comma. Count = 1002 - 999 = 3.
//     // This seems correct for the first comma position.
//     //
//     // k=2: power_of_1000 = 1000000, threshold = 999999.
//     // 1000000 > 1002, loop terminates.
//     //
//     // If `n = 10000`:
//     // k=1: power_of_1000 = 1000, threshold = 999.
//     // Count = 10000 - 999 = 9001.
//     // These are numbers from 1000 to 10000. All have one comma.
//     //
//     // k=2: power_of_1000 = 1000000, threshold = 999999.
//     // 1000000 > 10000, loop terminates.
//     // Total commas = 9001. This matches the manual calculation.
//
//     // The number of integers from `power_of_1000` to `n` inclusive is `n - power_of_1000 + 1`.
//     // But we are counting how many numbers *have* this comma.
//     // This means numbers that are at least `power_of_1000`.
//     // The count of numbers from 1 to `n` that are >= `power_of_1000` is `max(0LL, n - power_of_1000 + 1)`.
//     //
//     // Let's stick to the `n - threshold` logic for now, it seems to capture the count correctly.
//     // Example: n = 1002
//     // k=1: p=1000, t=999. count = 1002-999 = 3. (1000, 1001, 1002) -> 3 commas
//     //
//     // Example: n = 10000
//     // k=1: p=1000, t=999. count = 10000-999 = 9001. (1000...10000) -> 9001 commas
//     //
//     // Example: n = 1000000
//     // k=1: p=1000, t=999. count = 1000000 - 999 = 999001. (1000...1000000 have their 1st comma)
//     // k=2: p=1000000, t=999999. count = 1000000 - 999999 = 1. (Only 1000000 has its 2nd comma)
//     // Total commas = 999001 + 1 = 999002.
//     // Let's check this:
//     // Numbers 1000 to 999999: 999000 numbers. Each has 1 comma. Total 999000.
//     // Number 1000000: Has 2 commas.
//     // Total = 999000 + 2 = 999002. This matches.
//
//     // The count of numbers from `power_of_1000` to `n` is `n - power_of_1000 + 1`.
//     // These numbers each contribute *one* comma at this specific comma position.
//     // So, if `power_of_1000 <= n`, then all numbers from `power_of_1000` to `n`
//     // will have the comma corresponding to `power_of_1000`.
//     // The number of such integers is `n - power_of_1000 + 1`.
//     // Example: n = 1002. power_of_1000 = 1000. Count = 1002 - 1000 + 1 = 3. (1000, 1001, 1002). Add 3.
//     // Example: n = 10000. power_of_1000 = 1000. Count = 10000 - 1000 + 1 = 9001. (1000..10000). Add 9001.
//     // Example: n = 1000000.
//     // k=1: p=1000. Count = 1000000 - 1000 + 1 = 999001. Add 999001.
//     // k=2: p=1000000. Count = 1000000 - 1000000 + 1 = 1. Add 1.
//     // Total = 999001 + 1 = 999002.
//
//     // This seems correct. The count of numbers from `power_of_1000` to `n` inclusive.
//     // We only add if `n >= power_of_1000`. The loop condition `power_of_1000 <= n` ensures this.
//     total_commas += (n - power_of_1000 + 1);
//
//     // Prepare for the next comma position.
//     // We need to check for overflow when multiplying `power_of_1000` and `threshold`.
//     // `power_of_1000` can go up to 10^15. `power_of_1000 * 1000` could exceed `long long` limit (approx 9e18).
//     // 10^15 * 1000 = 10^18. This is within `long long`.
//     // The next `power_of_1000` would be 10^18. If `n` is 10^15, the loop will stop before exceeding.
//     // What if `n` is near 10^15?
//     // Example: n = 10^15.
//     // k=1: p=1000. Add 10^15 - 1000 + 1.
//     // k=2: p=10^6. Add 10^15 - 10^6 + 1.
//     // k=3: p=10^9. Add 10^15 - 10^9 + 1.
//     // k=4: p=10^12. Add 10^15 - 10^12 + 1.
//     // k=5: p=10^15. Add 10^15 - 10^15 + 1 = 1.
//     // k=6: p=10^18. 10^18 > 10^15. Loop ends.
//
//     // Check if `power_of_1000` will overflow if multiplied by 1000.
//     // The maximum value for `long long` is approximately 9e18.
//     // If `power_of_1000 > LLONG_MAX / 1000`, then `power_of_1000 * 1000` will overflow.
//     // If `n <= 10^15`, then `power_of_1000` will not exceed `10^15` during the loop.
//     // The next value `power_of_1000 * 1000` could be `10^18`. This fits in `long long`.
//     // The next value after that would be `10^21`, which overflows.
//     // Since `n <= 10^15`, we only need powers up to `10^15`.
//     // So `power_of_1000 * 1000` will not overflow for `n <= 10^15`.
//     if (power_of_1000 > n / 1000) { // Check to prevent overflow for the next `power_of_1000`
//         // If `power_of_1000 * 1000` would exceed `n`, we should stop.
//         // But the loop condition `power_of_1000 <= n` already handles this for the current iteration.
//         // We need to break if `power_of_1000 * 1000` would overflow or become larger than `n`.
//         // If `power_of_1000 > LLONG_MAX / 1000`, then `power_of_1000 * 1000` will overflow.
//         // Given `n <= 10^15`, the largest `power_of_1000` we care about is `10^15`.
//         // The next would be `10^18`.
//         // `power_of_1000` will not exceed `10^15` within the loop for `n <= 10^15`.
//         // So `power_of_1000 * 1000` will be at most `10^18`, which fits in `long long`.
//         // We can safely update `power_of_1000`.
//     }
//     power_of_1000 *= 1000;
//   }
//   return total_commas;
//
// Time Complexity:
// The loop iterates for powers of 1000 (10^3, 10^6, 10^9, ...) up to `n`.
// The number of iterations is logarithmic with respect to `n` in base 1000.
// For `n = 10^15`, the powers are 10^3, 10^6, 10^9, 10^12, 10^15. That's 5 iterations.
// So, the time complexity is O(log1000(n)), which is effectively O(log n).
//
// Space Complexity:
// We are using a few variables to store counts and powers of 10.
// The space used is constant, regardless of the input `n`.
// So, the space complexity is O(1).

class Solution {
public:
    long long countCommas(long long n) {
        // Base case: Numbers less than 1000 do not have any commas.
        if (n < 1000) {
            return 0;
        }

        long long total_commas = 0;
        // `power_of_1000` represents the start of ranges that have a comma at a specific position.
        // For example, 1000 for the first comma, 1000000 for the second, and so on.
        long long power_of_1000 = 1000;

        // Iterate through each comma position (thousands, millions, billions, etc.).
        // The loop continues as long as the current `power_of_1000` is less than or equal to `n`.
        // This ensures we only consider comma positions that can appear in numbers up to `n`.
        while (power_of_1000 <= n) {
            // `power_of_1000` marks the beginning of a new "comma group".
            // For example, if `power_of_1000` is 1000, we are counting numbers from 1000 onwards.
            // If `power_of_1000` is 1000000, we are counting numbers from 1000000 onwards.
            //
            // The number of integers from `power_of_1000` to `n` (inclusive) is `n - power_of_1000 + 1`.
            // Each of these numbers contributes *one* comma at this specific comma position.
            // For instance, numbers from 1000 to 1002 (if n=1002) all have their first comma.
            // The count is 1002 - 1000 + 1 = 3.
            //
            // We use `n - power_of_1000 + 1` because `power_of_1000` is the first number in the range that receives this comma.
            // E.g., for the first comma (thousands place), the numbers are 1000, 1001, ...
            // If n = 1002, the numbers are 1000, 1001, 1002. The count is 3.
            //
            // We add this count to `total_commas`.
            total_commas += (n - power_of_1000 + 1);

            // Prepare for the next comma position.
            // We need to check if multiplying `power_of_1000` by 1000 will exceed `n` or cause overflow.
            // Since `n` is up to 10^15, `power_of_1000` can go up to 10^15.
            // The next `power_of_1000` would be 10^18, which fits within `long long`.
            // If `power_of_1000 > n / 1000`, it implies that `power_of_1000 * 1000` would be greater than `n`.
            // In this case, the next iteration's `power_of_1000` will be greater than `n`, and the loop will terminate.
            // This check also helps prevent potential overflow if `n` was extremely close to `LLONG_MAX`.
            // Given `n <= 10^15`, `power_of_1000` will be at most `10^15`, and `power_of_1000 * 1000` will be at most `10^18`, which is safe for `long long`.
            // So, a simple multiplication is sufficient.
            if (power_of_1000 > n / 1000) { // Check to avoid overflow and to break early if the next power is already > n
                 break; // Exit loop if next `power_of_1000` would exceed `n` or overflow.
            }
            power_of_1000 *= 1000;
        }

        return total_commas;
    }
};
```

File: solutions/2026-09-09/python/2026-09-09-count-commas-in-range-ii.py
```python
# Problem: Count Commas in Range II
# Link: https://leetcode.com/problems/count-commas-in-range-ii/
#
# Approach:
# The problem asks us to count the total number of commas used when writing all integers from 1 to n.
# A comma is inserted after every three digits from the right. Numbers with fewer than 4 digits have no commas.
#
# The core idea is to realize that the number of commas for a number `x` depends on its number of digits.
# A number `x` will have `k` commas if `x` is greater than or equal to 1000, 1,000,000, 1,000,000,000, and so on.
# Specifically, a number `x` has:
# - 0 commas if 1 <= x <= 999
# - 1 comma if 1000 <= x <= 999,999
# - 2 commas if 1,000,000 <= x <= 999,999,999
# - 3 commas if 1,000,000,000 <= x <= 999,999,999,999
# - 4 commas if 1,000,000,000,000 <= x <= 999,999,999,999,999 (which is the maximum possible value for n)
#
# We can count the total number of commas by summing up the commas contributed by each "block" of numbers.
# For example, consider n = 1002.
# - Numbers from 1 to 999: 0 commas each. Total = 0.
# - Numbers from 1000 to 1002: each has 1 comma. Total = 3 * 1 = 3.
#
# Let's generalize. We can iterate through powers of 1000 (10^3, 10^6, 10^9, 10^12, ...).
# For a given power of 1000, say `p`, and the next power of 1000, `p_next`:
# - Numbers from `p` to `min(n, p_next - 1)` contribute one more comma than the numbers less than `p`.
#
# Let's define `count_commas(x)` as the total number of commas for numbers from 1 to `x`.
# We want to calculate `count_commas(n)`.
#
# `count_commas(x)`:
# - If `x < 1000`: return 0
# - If `1000 <= x < 10^6`: The numbers from 1000 to `x` each have 1 comma. The count is `x - 1000 + 1 = x - 999`.
# - If `10^6 <= x < 10^9`:
#   - Numbers from 1000 to 999,999: there are 999,000 such numbers, each with 1 comma. Total = 999,000.
#   - Numbers from 1,000,000 to `x`: each has 2 commas. This is where it gets tricky if we simply iterate.
#
# A better approach is to directly calculate the contribution of each magnitude.
#
# Let's define a function `calculate_commas_up_to(limit)` which calculates the total commas for numbers from 1 up to `limit`.
#
# For a `limit`:
#
# 1. Numbers from 1 to 999 contribute 0 commas.
# 2. Numbers from 1000 to 999,999:
#    - There are `999,999 - 1000 + 1 = 999,000` numbers in this range.
#    - Each of these numbers has exactly one comma.
#    - Contribution from this range: `min(limit, 999999) - 1000 + 1` (if `limit >= 1000`)
#    - Number of numbers with at least one comma: `max(0, limit - 999)`
#    - Total commas from this "thousands" group: `max(0, limit - 999)`
#
# 3. Numbers from 1,000,000 to 999,999,999:
#    - These numbers have at least two commas.
#    - The range is from `10^6` to `10^9 - 1`.
#    - Numbers with at least two commas are those >= `10^6`.
#    - Number of numbers with at least two commas: `max(0, limit - 999999)`. Each of these contributes an *additional* comma compared to the previous block.
#    - So, `max(0, limit - 999999)` additional commas.
#
# 4. Numbers from 1,000,000,000 to 999,999,999,999:
#    - These numbers have at least three commas.
#    - Numbers with at least three commas are those >= `10^9`.
#    - Number of numbers with at least three commas: `max(0, limit - 999999999)`. Each of these contributes yet *another* additional comma.
#    - So, `max(0, limit - 999999999)` additional commas.
#
# 5. Numbers from 1,000,000,000,000 to 999,999,999,999,999:
#    - These numbers have at least four commas.
#    - Numbers with at least four commas are those >= `10^12`.
#    - Number of numbers with at least four commas: `max(0, limit - 999999999999)`. Each of these contributes a fourth additional comma.
#    - So, `max(0, limit - 999999999999)` additional commas.
#
# The total number of commas for numbers from 1 to `limit` is the sum of these contributions:
#
# `total_commas = max(0, limit - 999) + max(0, limit - 999999) + max(0, limit - 999999999) + max(0, limit - 999999999999)`
#
# This formula calculates the total number of commas. For example, if `limit = 1002`:
# - `max(0, 1002 - 999) = max(0, 3) = 3`
# - `max(0, 1002 - 999999) = max(0, -998997) = 0`
# - `max(0, 1002 - 999999999) = max(0, ... ) = 0`
# - `max(0, 1002 - 999999999999) = max(0, ... ) = 0`
# Total = 3. Correct for Example 1.
#
# If `limit = 998`:
# - `max(0, 998 - 999) = max(0, -1) = 0`
# - `max(0, 998 - 999999) = 0`
# - `max(0, 998 - 999999999) = 0`
# - `max(0, 998 - 999999999999) = 0`
# Total = 0. Correct for Example 2.
#
# If `limit = 1000000`:
# - `max(0, 1000000 - 999) = 999001` (numbers 1000 to 1000000 have 1 comma each)
# - `max(0, 1000000 - 999999) = 1` (number 1000000 has an additional comma)
# - `max(0, 1000000 - 999999999) = 0`
# - `max(0, 1000000 - 999999999999) = 0`
# Total = 999001 + 1 = 999002.
# Let's verify:
# Numbers 1 to 999: 0 commas.
# Numbers 1000 to 999999: 999000 numbers, 1 comma each. Total = 999000.
# Number 1000000: 2 commas.
# Total for 1 to 1000000: 999000 (from 1000-999999) + 1 (from 1000000) from the first comma position + 1 (from 1000000) from the second comma position.
# This is becoming confusing. Let's re-evaluate the formula.
#
# The formula `max(0, limit - threshold) + ...` counts how many numbers are *greater than or equal to* the `threshold`.
#
# Let's think about the *number of commas inserted* at each "level" of thousands.
#
# Level 1 (thousands separator): applies to numbers >= 1000.
#   Number of integers in [1, n] that are >= 1000 is `max(0, n - 999)`.
#   These numbers contribute one comma *each* for this first separator.
#
# Level 2 (millions separator): applies to numbers >= 1,000,000.
#   Number of integers in [1, n] that are >= 1,000,000 is `max(0, n - 999999)`.
#   These numbers contribute a *second* comma *each*.
#
# Level 3 (billions separator): applies to numbers >= 1,000,000,000.
#   Number of integers in [1, n] that are >= 1,000,000,000 is `max(0, n - 999999999)`.
#   These numbers contribute a *third* comma *each*.
#
# Level 4 (trillions separator): applies to numbers >= 1,000,000,000,000.
#   Number of integers in [1, n] that are >= 1,000,000,000,000 is `max(0, n - 999999999999)`.
#   These numbers contribute a *fourth* comma *each*.
#
# Total commas = (commas from level 1) + (commas from level 2) + (commas from level 3) + (commas from level 4)
#
# Total commas = `max(0, n - 999)` + `max(0, n - 999999)` + `max(0, n - 999999999)` + `max(0, n - 999999999999)`
#
# This formula looks correct.
#
# For n = 10^15 (maximum constraint)
#
# `n - 999` is approximately 10^15
# `n - 999999` is approximately 10^15
# `n - 999999999` is approximately 10^15
# `n - 999999999999` is approximately 10^15
#
# Total commas will be roughly 4 * 10^15, which fits in a 64-bit integer (long long in C++, standard int in Python handles arbitrary precision).
#
# Time Complexity:
# The solution involves a constant number of arithmetic operations and `max` calls.
# The number of operations does not depend on the input `n` (it's fixed to 4 levels of commas).
# Therefore, the time complexity is O(1).
#
# Space Complexity:
# The solution uses a few variables to store intermediate results.
# The space used is constant and does not depend on the input `n`.
# Therefore, the space complexity is O(1).

class Solution:
    def countCommas(self, n: int) -> int:
        """
        Calculates the total number of commas used when writing all integers from 1 to n.
        """
        total_commas = 0

        # A comma is inserted after every three digits from the right.
        # Numbers with fewer than 4 digits contain no commas.

        # The first comma appears for numbers 1,000 and above.
        # Numbers >= 1000 have at least one comma.
        # The count of numbers in [1, n] that are >= 1000 is max(0, n - 999).
        # Each of these contributes one comma from the thousands separator.
        threshold1 = 1000
        # The number of integers from 1 to n that are >= threshold1
        count_ge_threshold1 = max(0, n - (threshold1 - 1))
        total_commas += count_ge_threshold1

        # The second comma appears for numbers 1,000,000 and above.
        # Numbers >= 1,000,000 have at least two commas.
        # The count of numbers in [1, n] that are >= 1,000,000 is max(0, n - 999999).
        # Each of these contributes a second comma from the millions separator.
        threshold2 = 1000000
        count_ge_threshold2 = max(0, n - (threshold2 - 1))
        total_commas += count_ge_threshold2

        # The third comma appears for numbers 1,000,000,000 and above.
        # Numbers >= 1,000,000,000 have at least three commas.
        # The count of numbers in [1, n] that are >= 1,000,000,000 is max(0, n - 999999999).
        # Each of these contributes a third comma from the billions separator.
        threshold3 = 1000000000
        count_ge_threshold3 = max(0, n - (threshold3 - 1))
        total_commas += count_ge_threshold3

        # The fourth comma appears for numbers 1,000,000,000,000 and above.
        # Numbers >= 1,000,000,000,000 have at least four commas.
        # The count of numbers in [1, n] that are >= 1,000,000,000,000 is max(0, n - 999999999999).
        # Each of these contributes a fourth comma from the trillions separator.
        threshold4 = 1000000000000
        count_ge_threshold4 = max(0, n - (threshold4 - 1))
        total_commas += count_ge_threshold4

        # Note: The constraint n <= 10^15 means we only need to consider up to 4 commas.
        # 10^12 is the smallest number with 4 commas (1,000,000,000,000).
        # 10^15 is the maximum input value.

        return total_commas

```